File: apps/server/src/services/action-queue.ts
import { addNotification, updateNotification } from './job-notifications';
import { logger } from '../utils/logger';

/**
 * Shared cross-user FIFO queue for server-side actions that do real Plex/API
 * work (playlist reorders, AI generation, cross-service imports, ...) so a
 * burst of requests from multiple users can't all run at once and so each
 * user can see how far down the queue their own request is. Bounded to a
 * small worker pool rather than one-at-a-time, since these actions are
 * independent of each other and don't need strict serialization - only a cap.
 *
 * In-memory only, not DB-persisted (unlike import-queue.ts's import_queue
 * table): these actions are seconds-to-low-minutes and trivially re-clickable
 * by the user, so losing a queued/in-flight one on a rare server restart is
 * an acceptable tradeoff - the same one job-notifications.ts already makes
 * for this class of transient progress.
 */

const MAX_CONCURRENT = 3;

interface QueuedAction {
  id: string;
  userId: number;
  handler: () => Promise<void>;
}

const queue: QueuedAction[] = [];
let active = 0;

/**
 * Queues an action and returns immediately with the caller's own position
 * (0 = about to run). `handler` does the real work and is responsible for
 * calling updateNotification(userId, notificationId, ...) itself to report
 * progress and the final success/error outcome, the same way every other
 * background job in this app already does.
 */
export function enqueueAction(
  userId: number,
  title: string,
  handler: (notificationId: string) => Promise<void>,
  notificationType: Parameters<typeof addNotification>[1]['type'] = 'action'
): { jobId: string; position: number } {
  const notification = addNotification(userId, {
    type: notificationType,
    title,
    detail: 'Queued',
    status: 'in-progress',
  });

  const job: QueuedAction = {
    id: notification.id,
    userId,
    handler: () => handler(notification.id),
  };
  queue.push(job);
  const position = queue.length - 1;

  refreshQueuedPositions();
  processQueue();

  return { jobId: job.id, position };
}

function refreshQueuedPositions(): void {
  queue.forEach((job, index) => {
    updateNotification(job.userId, job.id, {
      detail: index === 0 ? "Up next" : `Queued - position ${index + 1}`,
    });
  });
}

function processQueue(): void {
  while (active < MAX_CONCURRENT && queue.length > 0) {
    const job = queue.shift()!;
    active++;
    updateNotification(job.userId, job.id, { detail: 'Processing...' });
    refreshQueuedPositions();

    job.handler()
      .catch((error: any) => {
        // Handlers report their own success/error via updateNotification -
        // this only catches a handler that threw without doing so itself,
        // so the notification doesn't get stuck on "Processing..." forever.
        logger.error('Action queue job failed unexpectedly', { error: error.message, jobId: job.id, userId: job.userId });
        updateNotification(job.userId, job.id, { status: 'error', detail: error.message || 'Action failed' });
      })
      .finally(() => {
        active--;
        processQueue();
      });
  }
}

File: internal/services/actionqueue/actionqueue.go
// Package actionqueue ports services/action-queue.ts: a shared, cross-user
// FIFO queue for server-side actions that do real Plex/API work (playlist
// reorders, AI generation, cross-service imports, ...), bounded to a small
// worker pool so a burst of requests from multiple users can't all run at
// once, while each user can see how far down the queue their own request
// is. In-memory only - these actions are seconds-to-low-minutes and
// trivially re-clickable, so losing a queued/in-flight one on a rare
// restart is an acceptable tradeoff (same one the notifications package
// already makes).
package actionqueue

import (
	"log/slog"
	"strconv"
	"sync"

	"github.com/drevilish/playlist-lab/internal/services/notifications"
)

const maxConcurrent = 3

type job struct {
	id      string
	userID  int64
	handler func(notificationID string) error
}

type Queue struct {
	notifications *notifications.Store

	mu     sync.Mutex
	queued []job
	active int
}

func New(store *notifications.Store) *Queue {
	return &Queue{notifications: store}
}

// Enqueue queues an action and returns immediately with the caller's own
// position (0 = about to run). handler does the real work and is
// responsible for updating its own notification to report progress and
// the final outcome, the same way every other background job does.
func (q *Queue) Enqueue(userID int64, title string, notifType notifications.Type, handler func(notificationID string) error) (jobID string, position int) {
	n := q.notifications.Add(userID, notifType, title, "Queued", notifications.StatusInProgress, nil)
	j := job{id: n.ID, userID: userID, handler: handler}

	q.mu.Lock()
	q.queued = append(q.queued, j)
	position = len(q.queued) - 1
	q.refreshQueuedPositionsLocked()
	q.mu.Unlock()

	q.processQueue()
	return j.id, position
}

// refreshQueuedPositionsLocked must be called with q.mu held.
func (q *Queue) refreshQueuedPositionsLocked() {
	for i, j := range q.queued {
		detail := "Up next"
		if i > 0 {
			detail = "Queued - position " + strconv.Itoa(i+1)
		}
		q.notifications.Update(j.userID, j.id, notifications.Patch{Detail: &detail})
	}
}

func (q *Queue) processQueue() {
	for {
		q.mu.Lock()
		if q.active >= maxConcurrent || len(q.queued) == 0 {
			q.mu.Unlock()
			return
		}
		j := q.queued[0]
		q.queued = q.queued[1:]
		q.active++
		processing := "Processing..."
		q.notifications.Update(j.userID, j.id, notifications.Patch{Detail: &processing})
		q.refreshQueuedPositionsLocked()
		q.mu.Unlock()

		go q.run(j)
	}
}

func (q *Queue) run(j job) {
	defer func() {
		q.mu.Lock()
		q.active--
		q.mu.Unlock()
		q.processQueue()
	}()

	// Handlers report their own success/error via their notification - this
	// only catches one that returned an error (or panicked) without doing
	// so itself, so the notification doesn't get stuck on "Processing..."
	// forever.
	defer func() {
		if r := recover(); r != nil {
			slog.Error("action queue job panicked", "jobId", j.id, "userId", j.userID, "panic", r)
			status, detail := notifications.StatusError, "Action failed"
			q.notifications.Update(j.userID, j.id, notifications.Patch{Status: &status, Detail: &detail})
		}
	}()

	if err := j.handler(j.id); err != nil {
		slog.Error("action queue job failed unexpectedly", "jobId", j.id, "userId", j.userID, "error", err)
		status, detail := notifications.StatusError, err.Error()
		q.notifications.Update(j.userID, j.id, notifications.Patch{Status: &status, Detail: &detail})
	}
}
